### ISSUE TITLE
Re-tune the w_strategy="auto" heuristics after the plane-count and engine changes (R8, T6)

### ISSUE THREAD

**Issue by chrisfinlay** [chrisfinlay/jax-nufft#34]
Part of https://github.com/chrisfinlay/jax-nufft/issues/3 · Tracking: #35 · Depends on: #19, #17, #25, #26

## Context
`_auto_w_strategy_cpu/_gpu` gate on `n_w`, `w_kernel_width`, `window_padding_overhead` and `n_rows`; every one of
these changes after the width rule, nshift, w-flip, cost model, chunked strategy and windowed bucketing land. The
acceptance test replays the heuristic against `docs/benchmarks/v0.1.2-baseline-gpu.json`, which will be stale.

## Implementation plan
1. Re-run the GPU benchmark suite (`--runbench-gpu`) on the GH200 and the CPU sweep on a workstation; commit new
   baseline JSONs (`v0.2-baseline-*.json`).
2. Re-derive the gates (include the chunked strategy as a candidate; it likely replaces both `_scan` picks).
3. Update `tests/test_auto_strategy.py` (stub defaults and the "typical eps=1e-6" comment) and
   `test_auto_strategy_acceptance.py` to the new baseline.

Without GPU access: re-tune only `_auto_w_strategy_cpu` from a CPU sweep (five timing fixtures × all strategies,
timing protocol), leave `_auto_w_strategy_gpu` unchanged, and make `tests/test_auto_strategy_acceptance.py` skip
with a clear reason when the baseline JSON's rows lack the new plan fields (the heuristic's inputs changed); note the
pending GPU re-tune in the PR.

## Definition of done
- [ ] Acceptance test: auto pick within 15 % of the best measured strategy on every cell of the new baseline (GPU), or the
      CPU-only variant above with the GPU re-tune explicitly deferred.
- [ ] README "auto" section updated.

## Working context (self-contained)

- **Environment.** `pixi run -e test <cmd>` from the repo root (Python 3.12; jax, jax-finufft, numpy; ducc0 as the test
  oracle). Read `AGENTS.md` §1–§6 first: sign convention, plan-then-call pattern and the plan-field checklist (§4),
  strategies (§5), test/benchmark commands (§6). `pixi run -e test pytest -q --runslow`, `pixi run -e dev lint` and
  `pixi run -e dev typecheck` must stay green.
- **Fixtures.** `tests/conftest.py` defines telescopes `EDA2` (64², 120° FoV, 400 rows), `MWA_COMPACT` (128²),
  `MWA_EXTENDED` (256²), `MEERKAT` (256²), `GH200_LARGE` (2048², 50k rows) and `synthetic_uvw(tel, zenith_angle_deg, seed)`.
  "`<name>` zenith / off30" means `uvw = synthetic_uvw(tel, 0.0 or 30.0, seed=0)`, `freq = np.array([tel.freq_hz])`,
  `image_shape = (tel.n_pix, tel.n_pix)`, `pixsize = tel.pixsize`, image `np.random.default_rng(7).standard_normal(...)`
  and complex vis from the same generator. The "seven review fixtures" are EDA2 zenith, MWA_compact zenith and off30,
  MWA_extended zenith and off30, MeerKAT zenith and off30. The "five timing fixtures" are MWA_compact off30,
  MWA_extended off30, MeerKAT off30, EDA2 zenith, MWA_extended zenith.
- **Accuracy metric.** Relative L2 error `‖a − b‖₂ / ‖b‖₂` against the exact DFT in the repo's sign convention
  (`tests/test_against_dft.py::_reference_forward`, `tests/test_adjoint.py::_reference_adjoint`; vectorise over row
  chunks to make 256² fixtures take seconds). "x× eps" is that ratio divided by the requested epsilon. "The accuracy
  sweep" is the seven review fixtures × eps ∈ {1e-3, 1e-4, 1e-5, 1e-6, 1e-7, 1e-8, 1e-10, 1e-12}, forward and adjoint,
  `dense_scan` (the other strategies agree with it to 1e-14). The width-rule sub-issue adds
  `tests/test_accuracy_sweep.py` (gated by `--runsweep`) that produces this table; later issues re-run it.
- **ducc0 as oracle.** `ducc0.wgridder.dirty2vis(uvw=, freq=, dirty=, pixsize_x=, pixsize_y=, epsilon=, do_wgridding=True,
  divide_by_n=False, nthreads=1)` and `ducc0.wgridder.vis2dirty(..., npix_x=, npix_y=, divide_by_n=True)` match
  jax-nufft's default conventions (see `tests/test_against_ducc.py`). With `verbosity=1` ducc0 prints its plane count to
  C-level stdout as `(nu x nv x nplanes), supp=W`; capture it by redirecting file descriptor 1 (`os.dup2` to a temp file,
  not `contextlib.redirect_stdout`) and parse `r"\((\d+)x(\d+)x(\d+)\), supp=(\d+)"`.
- **Timing protocol.** Same machine before and after; `nthreads` stated; plan built and one warm-up call done outside the
  timer; median of ≥ 7 calls, each ending with `.block_until_ready()`. Timing gates are relative to the pre-PR number
  measured this way; absolute times quoted above are from the review machine (Apple M-series, 10 cores) for orientation only.
- **Memory protocol.** `jax.jit(fn).lower(*args).compile().memory_analysis().temp_size_in_bytes` (works on the CPU backend);
  compare with the same quantity for the plain forward/adjoint call.
- **GPU gates.** Need the CUDA jax-finufft build (pixi env `gpu-dev`, Linux). Without a GPU: satisfy the CPU gates, state in
  the PR that the GPU gate is pending for the maintainer, and use `docs/benchmarks/v0.1.2-baseline-gpu.json` (GH200
  per-cell median seconds, `n_w`, padding overhead) as the reference numbers.
- **PR conventions.** One PR per issue, branched from `dev`, title `<short> (closes #N)`, the definition-of-done checklist
  copied into the description with before/after numbers, and the statement that ducc0 source was not consulted.

Review finding IDs (private review document): R8, T6.

---
**Sources to implement from.** ducc0 is GPL-2.0-or-later; it is used in this project only as a black-box test
oracle through its public Python API. Do not consult or transliterate its source for this work; implement from the
mathematics stated above and these references (FINUFFT and jax-finufft are Apache-2.0 and may be read freely):

- [Arras+2021] P. Arras, M. Reinecke, R. Westermann, T. A. Enßlin, *Efficient wide-field radio interferometry response*, A&A 646, A58 (2021), arXiv:2010.10122 — §2 w-stacking, eq. 12 plane spacing, eq. 19 Hermitian symmetry, eqs. 20–21 plane placement, eq. 29 modified ES kernel, §3.2 + Appendix A optimal kernel parameters, §4.1 cost model.
- [Barnett+2019] A. H. Barnett, J. Magland, L. af Klinteberg, *A parallel nonuniform fast Fourier transform library based on an "exponential of semicircle" kernel*, SIAM J. Sci. Comput. 41(5), C479 (2019), arXiv:1808.06736 — eq. 8 ES kernel, eq. 10 β = 2.30·W and the width rule, Theorem 7 error decay, eq. 18 Gauss–Legendre kernel transform.
- [Ye+2022] H. Ye, S. F. Gull, S. M. Tan, B. Nikolic, *Optimal gridding and degridding in radio interferometry imaging*, MNRAS 510, 4110 (2022), arXiv:2110.03914 — kernel optimisation methodology.
- [TMS2017] A. R. Thompson, J. M. Moran, G. W. Swenson, *Interferometry and Synthesis in Radio Astronomy*, 3rd ed., Springer (2017), Ch. 3 — conjugate symmetry of the visibility function and the shift theorem.
- [Higham2002] N. J. Higham, *Accuracy and Stability of Numerical Algorithms*, 2nd ed., SIAM (2002), Ch. 1 — cancellation and range reduction.
- [FINUFFT] `src/spreadinterp.cpp::setup_spreader` (Apache-2.0) — the implemented width rule `W = ceil(-log10(eps/10))` at σ = 2.
- [jax-finufft] `lib/jax_finufft_cpu.cc` (Apache-2.0) — one `makeplan` per custom call, one `setpts` per point set.

**Evidence.** The measurements quoted above are from the 2026-09-02 review run (jax 0.9.2, jax-finufft 1.3.1,
ducc0 0.41.0 as oracle, macOS arm64 10-core, float64 unless stated). The review document is kept outside the
repository and is not to be committed; the definition-of-done tests below are the reproduction.

**Comment by chrisfinlay** [comment]
Measured evidence that the **default `w_strategy` is the wrong default on GPU**, and that `auto` already picks
correctly — so this may be as much about changing the default as about re-tuning the heuristic.

Measured on one NVIDIA GH200 (Daint, nid006546, driver 590.48.01), against ducc0 on 72 Grace cores of the same node,
eps = 1e-6, float64, single channel. jax-nufft and ducc0 agree to 7.5e-7 (0.75x eps), so this is like-for-like.
Forward / adjoint milliseconds, median of 9 with warm-up outside the timer:

| fixture | n_w | ducc0 (72 cores) | jax `dense_scan` (**the default**) | jax `dense_vmap` (what `auto` picks) |
|---|---|---|---|---|
| MWA_extended off30, 256^2, 600 rows | 251 | 89.8 / 91.1 | **501.9 / 331.8** | 30.9 / 14.4 |
| MeerKAT off30, 256^2, 600 rows | 19 | 17.2 / 17.9 | 39.4 / 26.2 | 4.9 / 3.3 |
| GH200_large off30, 2048^2, 50k rows | 43 | 67.9 / 184.8 | 160.5 / 159.2 | 47.6 / 78.8 |

So on GPU:

* with `w_strategy="dense_vmap"` (or `"auto"`), jax-nufft is **1.4-3.5x faster than ducc0 on the forward and
  2.3-6.3x on the adjoint**;
* with the shipped default, it is **1.4-5.6x slower than ducc0** — 501.9 ms versus 89.8 ms on MWA_extended off30.

A user who takes the defaults on GPU therefore gets the worst available configuration on hardware where the right
one comfortably beats a 72-core CPU.

`auto` resolves correctly on this hardware, checked directly:

```
MWA_extended_off30     n_w=251   auto_fwd=dense_vmap     auto_adj=dense_vmap
MeerKAT_off30          n_w=19    auto_fwd=dense_vmap     auto_adj=dense_vmap
GH200_large_off30      n_w=43    auto_fwd=dense_vmap     auto_adj=windowed_vmap
```

Two suggestions for this issue's scope:

1. **Consider making `"auto"` the default `w_strategy`**, rather than only re-tuning its thresholds. The heuristic
   is already right here; nothing reaches it. If the default stays `dense_scan` for CPU-side reasons, the platform
   check could at least select `auto` when the backend is GPU. Whatever is decided, the reasoning belongs in the
   docstring, because the current default is a silent 5.6x penalty on GPU.
2. **Re-tune against these numbers.** The GPU advantage is not uniform: it is largest where there are many planes to
   batch (MWA_extended off30, 251 planes, 6.3x on the adjoint) and smallest on the large NUFFT-dominated fixture
   (GH200_large, 43 planes and 50k rows, 1.4x on the forward). Any re-tuned heuristic should be checked against both
   regimes, not just one.

Also worth recording, since it mirrors #24: ducc0 has the same small-problem threading pathology we fixed. On
MWA_extended off30 it takes 89.8 ms at 72 threads but **74.1 ms single-threaded**. Comparisons against ducc0 should
state its thread count, and should not assume more threads is faster.

Reproducer: `/users/cjfinlay/pasc/chris/wgridder-gpu/` on Daint holds `released/` and `branch/` checkouts with the
`gpu-dev` environment, `vs_ducc.py`, `auto_check.py` and the raw results. Run with
`srun -A c39 -p debug -N1 -n1 -c 72 --gpus-per-task=1`; note that without `-c` Slurm pins the task to a single core
and ducc0 is silently crippled.

**Comment by chrisfinlay** [comment]
## Evidence from #43: the CPU padding gate mis-selects on ~83% of random draws

While correcting `window_padding_overhead`'s denominator (#43) I measured how stable the
`_CPU_PADDING_CUTOFF` decision is across `synthetic_uvw` seeds. It is not stable, and the
instability is concentrated on the one fixture AGENTS.md section 9 records as the measured
windowed **win** — MWA_extended off-zenith, eps 1e-3, adjoint.

The whole calibration grid is measured at `seed=0`. That seed is an outlier at the *bottom* of
the distribution:

| seed | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 |
|---|--:|--:|--:|--:|--:|--:|--:|--:|--:|--:|--:|--:|
| overhead | **5.78** | 6.27 | 6.32 | 7.04 | **5.85** | 6.61 | 6.33 | **8.09** | 6.18 | 6.43 | 6.56 | 6.88 |
| adjoint pick | windowed | dense | dense | dense | windowed | dense | dense | dense | dense | dense | dense | dense |

Range 5.78–8.09, median 6.38. The gate fires on **10 of 12 seeds**, sending `auto` to
`dense_scan` on a configuration where the windowed adjoint is the measured CPU win. Seed 0 and
seed 4 are the only two that land on the fast path.

**This is pre-existing, not something #43 introduced.** On the old (padded) metric against the
old 5.0 cutoff the same seeds fire 10 of 12 as well, and the decision is identical seed-for-seed:

| | fires | decisions changed by #43 |
|---|--:|--:|
| old metric, cutoff 5.0 | 10 / 12 | — |
| corrected metric, cutoff 6.0 | 10 / 12 | **0 / 12** |

So the 5.0 → 6.0 restatement tracks the metric's ~20% correction exactly and preserves the gate
behaviour; what it also does is make the fragility legible, because the corrected number is the
one with a physical meaning.

## Why an absolute cutoff is the wrong shape here

The regime this fires in is precisely where windowing is cheapest. What decides
windowed-vs-dense is `max_window_size / n_rows`, which on the corrected definition is
`window_padding_overhead · w_kernel_width / n_w` up to edge effects. On MWA_extended off30 at
`n_w/W = 62`, that ratio is ~0.09 — the windowed path touches about a tenth of the rows the
dense path does. A high padding overhead and a large `n_w / w_kernel_width` is a windowed
**win**, and the absolute cutoff reads only the first of those two.

PR #44 investigated substituting the ratio directly and found it reproduces all 40 calibration
decisions but puts three within 5% of the boundary, which is why it was not adopted there. That
finding stands, but it was measured on `seed=0` alone — given the spread above, "within 5% of
the boundary at one seed" is not the objection it looked like.

Suggested for this issue: re-tune against a seed *distribution* rather than a single draw, and
gate on `max_window_size / n_rows` (or the equivalent `overhead · W / n_w`) rather than on the
overhead alone. Both need CPU timings on the far side of the cutoff, which this issue already
scopes.
